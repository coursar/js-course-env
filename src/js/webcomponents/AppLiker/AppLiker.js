const template = document.createElement('template');
template.innerHTML = `
<div>
  <span data-id="likes">0</span>
  <button data-action="like">+</button>
  <button data-action="dislike">-</button>
</div>
`;

class AppLiker extends HTMLElement {
  static get observedAttributes() {
    return ['prop'];
  }

  #likesEl;

  #likeButtonEl;

  #dislikeButtonEl;

  #likes = 0;

  constructor() {
    super();

    const shadow = this.attachShadow({ mode: 'open' });
    const tpl = template.content.cloneNode(true);

    this.#likesEl = tpl.querySelector('[data-id="likes"]');
    this.#likeButtonEl = tpl.querySelector('[data-action="like"]');
    this.#dislikeButtonEl = tpl.querySelector('[data-action="dislike"]');

    // this.onLike = this.onLike.bind(this);
    this.#likeButtonEl.addEventListener('click', () => this.onLike());

    shadow.appendChild(tpl);
  }

  onLike() {
    console.log(this);
    this.#likes += 1;
    this.#updateView();
  }

  set likes(val) {
    this.#likes = val;
    this.#updateView();
  }

  get likes() {
    return this.#likes;
  }

  #updateView() {
    this.#likesEl.textContent = `${this.#likes}`;
  }

  connectedCallback() {
    console.log('connected');
  }

  disconnectedCallback() {
    console.log('disconnected');
  }

  attributeChangedCallback(name, oldVal, newVal) {
    console.log('attr changed', name, oldVal, newVal);
  }
}

customElements.define('app-liker', AppLiker);

export default AppLiker;
