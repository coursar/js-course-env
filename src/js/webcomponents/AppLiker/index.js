import AppLiker from './AppLiker';

export default AppLiker;
