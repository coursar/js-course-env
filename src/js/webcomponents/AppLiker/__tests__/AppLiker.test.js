/**
 * @jest-environment jsdom
 */

import '../AppLiker';

test('should render liker', () => {
  const containerEl = document.createElement('div');
  const likerEl = document.createElement('app-liker');
  containerEl.appendChild(likerEl);
  expect(likerEl.shadowRoot.children).toMatchSnapshot();

  likerEl.shadowRoot.querySelector('button[data-action="like"]').click();
  expect(likerEl.likes).toBe(1);
});
