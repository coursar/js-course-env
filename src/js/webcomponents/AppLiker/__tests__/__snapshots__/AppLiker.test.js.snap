// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render liker 1`] = `
HTMLCollection [
  <div>
    
  
    <span
      data-id="likes"
    >
      0
    </span>
    
  
    <button
      data-action="like"
    >
      +
    </button>
    
  
    <button
      data-action="dislike"
    >
      -
    </button>
    

  </div>,
]
`;
